Add an afterburner run as a target

NGoetz opened the pull request:
This PR introduces a new target for the analysis suite. It runs 20 times SMASH as an afterburner for a sampled particle list at 200 GeV and 5 TeV. The results are analyzed using the present functionalities for the energy scan target. Similar report pages are created as well. The comparison to previous results is included in this PR. The generated report can be seen [here](https://vfs.fias.science/d/e401040a37/).

It would be helpful if the review checks again that I did not break any functionality of the energy scan target, as I had to modify there some scripts to reuse them.

gabriele-inghirami commented on test/afterburner/afterburner.yaml:
```diff
@@ -0,0 +1,21 @@
+
+Output:
+    Particles:
+        Format:          ["Binary", "Oscar2013"]
```
Why do we need also the Oscar2013 output?

NGoetz replied:
You are right, this is not necessary. I removed Oscar output.

gabriele-inghirami commented on test/energy_scan/mult_and_spectra_plotter.py:
```diff
@@ -336,17 +346,20 @@ def is_number(s):
-
     for quantity in quantities:
         if ('spectra' in quantity): continue
+        #if ('afterburner' in colliding_systems): continue
```
If not needed anymore, perhaps it can be canceled instead of just being commented out.

NGoetz replied:
Thank you, I forgot to remove a few stray comments.

gabriele-inghirami commented on test/afterburner/CMakeLists.txt:
```diff
@@ -0,0 +1,200 @@
+    set(smash_afterburner_config "${CMAKE_CURRENT_SOURCE_DIR}/afterburner.yaml")
+
+    execute_process (
+    COMMAND bash -c "cd ${CMAKE_CURRENT_SOURCE_DIR}/${test_name} && unzip -o sampling0.zip"
```
Since this is a single compressed file and not an archive, I am wondering it it would be better to use gzip...
Currently in the Virgo cluster we are using containers, so we can put inside what we want, moreover `zip` usually is very common elsewhere, however it is not installed in the "bare metal" OS:
```
g@uhuru:~/SMASH-dev/smash-analysis/test/afterburner$ ssh -J lxpool ginghir@virgo.hpc.gsi.de
X11 forwarding request failed on channel 0
Virgo Slurm Cluster – Job Submit Node
ginghir@lxbk0599:~$ zip -v
bash: zip: command not found
```
gzip at the highest compression level performs slightly better than standard zip:
```
g@uhuru:~/SMASH-dev/smash-analysis/test/afterburner/LHC$ gzip -9 sampling0
g@uhuru:~/SMASH-dev/smash-analysis/test/afterburner/LHC$ ls -l
total 175644
-rw-rw-r-- 1 g g 89696892 Dec 14 13:37 sampling0.gz
-rw-rw-r-- 1 g g 90156516 Mar 30 10:12 sampling0.zip
```
However, I think that it's unlikely that this will ever be an issue...

NGoetz replied:
I was not aware of this problem, thank you! I changed it to gzip

gabriele-inghirami commented:
Overall, the changes look fine.
However, the scripts are rather convoluted, they handle values of variables read/imported from other files and it is hard (at least for me) to tell if everything is OK just by inspecting the code.
As suggest by Niklas, one should probably run both the `energy_scan` and the `afterburner` targets.
I can start these runs today, if @justinmohs is not already taking care of that.

justinmohs commented:
> Overall, the changes look fine.
> However, the scripts are rather convoluted, they handle values of variables read/imported from other files and it is hard (at least for me) to tell if everything is OK just by inspecting the code.
> As suggest by Niklas, one should probably run both the `energy_scan` and the `afterburner` targets.
> I can start these runs today, if @justinmohs is not already taking care of that.

It would be nice if you could tests the targets @gabriele-inghiramim since I have never installed the hybrid or root and it would take me quite some time to do so. I can still have another look at the code if this is still useful.

NGoetz commented:
Thank you @gabriele-inghirami for the comments, I modify this points today.
@justinmohs this is only an afterburner run, so neither hybrid nor root are required to be installed. 

I did a test run of both energy scan and afterburner target myself but I think due to the size of the change it is better if someone else double checks that nothing is broken.

justinmohs commented:
Ah true, then I will go ahead and try it on my own

gabriele-inghirami commented:
>  then I will go ahead and try it on my own

OK, so, you do the tests and I can stop here, right? (just to avoid either do both the same thing or both nothing)

justinmohs commented:
On Mac I now get this warning when running cmake:
<pre>
Apple gzip 352.40.6
usage: gunzip [-123456789acdfhklLNnqrtVv] [-S .suffix] [<file> [<file> ...]]
 -1 --fast            fastest (worst) compression
 -2 .. -8             set compression level
 -9 --best            best (slowest) compression
 -c --stdout          write to stdout, keep original files
    --to-stdout
 -d --decompress      uncompress files
    --uncompress
 -f --force           force overwriting & compress links
 -h --help            display this help
 -k --keep            don't delete input files during operation
 -l --list            list compressed file contents
 -N --name            save or restore original file name and time stamp
 -n --no-name         don't save original file name or time stamp
 -q --quiet           output no warnings
 -r --recursive       recursively compress files in directories
 -S .suf              use suffix .suf instead of .gz
    --suffix .suf
 -t --test            test compressed file
 -V --version         display program version
 -v --verbose         print extra statistics
-- Configuring afterburner/LHC.
gunzip: invalid option -- o
Apple gzip 352.40.6
usage: gunzip [-123456789acdfhklLNnqrtVv] [-S .suffix] [<file> [<file> ...]]
 -1 --fast            fastest (worst) compression
 -2 .. -8             set compression level
 -9 --best            best (slowest) compression
 -c --stdout          write to stdout, keep original files
    --to-stdout
 -d --decompress      uncompress files
    --uncompress
 -f --force           force overwriting & compress links
 -h --help            display this help
 -k --keep            don't delete input files during operation
 -l --list            list compressed file contents
 -N --name            save or restore original file name and time stamp
 -n --no-name         don't save original file name or time stamp
 -q --quiet           output no warnings
 -r --recursive       recursively compress files in directories
 -S .suf              use suffix .suf instead of .gz
    --suffix .suf
 -t --test            test compressed file
 -V --version         display program version
 -v --verbose         print extra statistics
-- Configuring done
-- Generating done
</pre>

NGoetz commented:
Sorry I forgot to remove an option in the latest commit. Now it should work.

gabriele-inghirami commented:
By replacing two big binary files, now we have increased the git archive by 130 MB, i.e. by roughly 30%, because git does not forget the old commits. Hopefully, the problem should disappear after the merging, when you will delete your branch.
However, once in master, the data load will remain there "forever".
If, in the future, other replacements are foreseen, we should probably either plan a pruning procedure (feasible, but not easy and with several drawbacks, see e.g.: https://stackoverflow.com/questions/18010048/how-to-permanently-delete-a-commit-from-gits-history ) or adopt another strategy (e.g. pull the data from an online archive or
https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository )
@AxelKrypton @stdnmr : what do you think?
Perhaps, also @elfnerhannah has comments...
Of course, if we don't care about the future size of the archive or if large binary data files will remain an exceptional event, then everything is fine.

NGoetz commented:
Could this be fixed by squashing commits?

AxelKrypton commented:
> @AxelKrypton @stdnmr : what do you think?

I just cloned the master and the repository is over 360MB. I was surprised and reading here I see that there are pretty large binary files in the repository. Nowadays few hundreds MB are not insane, but if this is going to be on the GB scale soon, I would rather think of an alternative, if any is available. Decoupling big data from the repository sounds good to me, especially if the responsibility of this repository is not to contain data-sets. We should discuss this further all together.

justinmohs commented:
currently, when running cmake (in my case for the second time) I get the following:

<pre>
sampling0 already exists -- do you wish to overwrite (y or n)? y
-- Configuring afterburner/LHC.
sampling0 already exists -- do you wish to overwrite (y or n)? y
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/justinmohs/smash-analysis/build
justinmohs@client84-on-sette build % git status
On branch ngoetz/afterburner_run
Your branch is up to date with 'origin/ngoetz/afterburner_run'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    ../test/afterburner/LHC/sampling0.gz
	deleted:    ../test/afterburner/RHIC/sampling0.gz

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	../test/afterburner/LHC/sampling0
	../test/afterburner/RHIC/sampling0

no changes added to commit (use "git add" and/or "git commit -a")
</pre>
I think the user being asked whether the sampling file should be replaced is fine (even though I was a little confused in the beginning). I would prefer not to delete the `sampling0.gz` file though, because even though one only wants the unzipped file, I don't like changing or deleting files tracked by git.

NGoetz commented:
I have added the correct gzip options to not remove the tracked files.
Normally no big binary files are required in this repository. For an afterburner run, however, the sampled particle list is needed as an input. If this is an issue, we could consider storing them in a different way, outside of the repository, too.

stdnmr commented:
> Could this be fixed by squashing commits?

Not sure if I follow the issue completely, but if the problem is that the repo size is enlarged by commiting the required binary files multiple times, squashing could help, but one needs to check this in advance. Since this is not merged yet, one could also easily remove past binary files from the history **on this branch**.

gabriele-inghirami commented:
> This must be done without rewriting any past history that is already on any other branch, e.g. master.

Good point, but I think that at the moment we should be in a good position for doing this operation with damaging anything...

elfnerhannah commented:
I am certainly in favour of not storing the data files in the git repository! This was my first comment when I started looking into this branch that this looked funky to me.

NGoetz commented:
I could put the input data in an online archive and download it during compilation. After that change, I can remove it from the git history.
Which archive should we use for this?

gabriele-inghirami commented:
> Which archive should we use for this?

Maybe a directory within the web space at GSI? 
However, I would suggest to not include the online archive address in the code, but pass it as an argument in slurm-job.sh, together with the other variables: it will be easier to maintain and more flexible for us.
Of course this should be explained somewhere in the documentation.

justinmohs commented on test/afterburner/CMakeLists.txt:
```diff
@@ -0,0 +1,163 @@
+  add_custom_command(OUTPUT ${plot}
+      COMMAND "python3" "${CMAKE_SOURCE_DIR}/test/energy_scan/mult_and_spectra_plotter.py"
+              "--theory" ${all_analysis_outputs}
+              "--config" "${CMAKE_CURRENT_BINARY_DIR}/RHIC/data/config.yaml"
```
I think there should be a `/RHIC/data/1/config.yaml`

NGoetz replied:
Thank you, I will fix this.

justinmohs commented:
I am not sure if this s really significant or at all related to your implementation but it may be that the prefactors are not properly taken into account for the data from the old smash version. See the AGS Au+Au pt spectra of protons for example:[ptspectra_AuAuPbPb_AGS_2212.pdf](https://github.com/smash-transport/smash-analysis/files/8384587/ptspectra_AuAuPbPb_AGS_2212.pdf)
Also unrelated to your work, I had a really tough time to get the analysis suite to run because the keyword `nonposy` is not available in the newest matplotlib version (see https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.pyplot.semilogy.html compared to https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.semilogy.html).

justinmohs commented:
A minor thing I noticed was that the `target_name` is not defined when running the `afterburner_plots` target. The only consequence seems to be that one gets the output `Plotting results for .`. That might be a relatively quick fix.

NGoetz commented:
I have fixed the missing 1.

Both the scaling of former results and the problem with posy are present on master too, we should address this in future. However, the scaling problem is only happening for plots which are not part of our report, so while definitely a bug, it is not of high urgency.
Regarding the plotting target, I modified it a bit, because there should be no separate plotting runs as everything is plotted together. Now there is only on single plotting target.
I have also updated the README according to my changes.

justinmohs commented:
I tried the afterburner and energy scan target again and both worked fine. We could think of adding the link to the sampled lists to the README but having the link, everything was clear. 
Since you had a closer look at the code than me, maybe you could approve the PR @gabriele-inghirami.

gabriele-inghirami commented:
> maybe you could approve the PR

I think that you could do it, as well. :)
However, OK, I will take the blame if something not working properly comes out later...

gabriele-inghirami reviewed (approved):
Overall the code seems to work fine. Possible future minor improvements are mentioned in the comments.
Possible issues might still emerge during the real production use, as with any software, but right now I don't see any valid reason to postpone the approval of this PR.
Thank you! @NGoetz @justinmohs
